forem/forem
# Add ability to lock discussions

## What type of PR is this? (check all applicable)

## Description
This PR introduces functionality to allow **article authors, admins, and super admins** to lock discussions which basically prevents the ability for users to add any further comments to an article. We've also added some logic to allow a permitted user to include a reason that will display publicly when a discussion is locked.

**I'm looking for feedback on this approach while I work through the remainder of the TODO list - this PR is _not_ ready yet.**

**TODO**
- [ ] Frontend tests (😭 )
- [x] ~Add CRUD functionality for discussion locks to Admin~ (~potentially~ a separate PR)
- [ ] Update/test error handling
- [ ] **Polish design (could use feedback here as well)**
- [x] Bust article cache after creating/destroying a discussion lock.
- [ ] **Fix keyboard navigation in the confirmation modal.**

A few things to note...
- **The JavaScript mess** - because we cache articles (somewhat aggressively) at the edge, adding "dynamic" functionality to articles becomes a bit of a pain. For example, once we lock a discussion, we'll bust the cache for the article which will update the markup. However, this is done asynchronously so we basically have to take care of all the updates in JS which is why we're always showing the markup and using JS to hide/show appropriately.
- **Naming** - we have different naming conventions flying around (comments, threads, discussions, etc.). We ultimately landed on **discussion** because 1) that's what we show in the UI and 2) that's what the customer success team liked the most.
- **Architecture** - for now, discussion locks are limited to Articles. We set this up in a way that allows it to be extensible in the future. We could make the `DiscussionLock` table and relationship polymorphic, write a `DataUpdateScript`, and then `DiscussionLocks` could be added to other domains in the future if needed. We also didn't want to further "crowd" the `Article` domain as it's quite large already.
- **MVP** - this was designed under the premise of a "quick win" more than it was a "perfect solution". There are many areas for improvement in the future. Things like allowing admins to assign any user they choose the ability to lock discussions, updating the UX, etc. We spoke with the customer success team to balance the "ideal" with the "quick and practical". If you go to the linked issue, you'll see we never really decided on a design, but we used that as a guide.
- **Single PR** - you may be wondering why this wasn't split into multiple PRs - great question. We've done that in the past and that approach carries some pros and cons. Splitting it makes reviews more "consumable", but it often leads to repetition of context that's lost between PRs. By keeping it in one PR, all the context is there and it can be easier to understand what's going on with everything in front of you. When starting this, we felt this feature would be right on the cusp of too big for a single PR.

## Related Tickets & Documents
closes https://github.com/forem/forem/issues/9619

## QA Instructions, Screenshots, Recordings
https://user-images.githubusercontent.com/15987080/118851863-36498c80-b8a0-11eb-816d-84e2c7a33f33.mp4

1. Pull down this branch and run migrations.
2. Fire up the app locally and sign in.
3. Write and publish a post.
4. Fire up your browser's dev tools/console and keep an eye on it over the remaining steps for any related errors. You'll likely see some Pusher and HoneyBadger errors - that's normal.
5. Lock the discussion on the post.
6. Unlock the discussion on the post.
7. Lock the discussion on the post again.
8. Unlock the discussion on the post again.

### UI accessibility concerns?
**I need help getting keyboard navigation to work on the modal.**

## Added tests?

## [Forem core team only] How will this change be communicated?
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post

![wedding_crashers_lock_it_up_gif](https://media.giphy.com/media/3ohzdFvSAgvmXq6n2o/giphy.gif)
